File: app/src/main/java/com/sleticalboy/dailywork/PermissionCheckActivity.java
package com.sleticalboy.dailywork;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

/**
 * Created on 18-1-31.
 *
 * @author sleticalboy
 * @version 1.0
 * @description
 */
public class PermissionCheckActivity extends AppCompatActivity {

    protected String[] permissions = {
            Manifest.permission.READ_CONTACTS,
            Manifest.permission.WRITE_CONTACTS,
            Manifest.permission.READ_PHONE_STATE,
    };
    protected static final int REQUEST_CODE = 1000;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            checkPermissions();
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    protected void checkPermissions() {
        if (ContextCompat.checkSelfPermission(this, permissions[0]) != 0
                || ContextCompat.checkSelfPermission(this, permissions[1]) != 0) {
            ActivityCompat.requestPermissions(this, permissions, REQUEST_CODE);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CODE && grantResults.length > 0) {
            for (int i = 0; i < grantResults.length; i++) {
                if (grantResults[i] != PackageManager.PERMISSION_GRANTED) {
                    Log.d("PermissionCheckActivity", permissions[i] + " 没有授权");
                }
            }
        }
    }
}

File: MainApp/src/main/java/com/sleticalboy/dailywork/base/PermissionCheckActivity.kt
package com.sleticalboy.dailywork.base

import android.Manifest
import android.annotation.TargetApi
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AppCompatActivity
import android.util.Log

/**
 * Created on 18-1-31.
 *
 * @author sleticalboy
 * @version 1.0
 * @description
 */
open class PermissionCheckActivity : AppCompatActivity() {

    /**
     * 必须要动态授权的权限, 每个权限代表一个权限组, 只要授权一个, 该组的权限都可以使用
     */
    private val permissions = arrayOf(
            Manifest.permission.READ_CONTACTS,
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.READ_PHONE_STATE
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            checkPermissions()
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    private fun checkPermissions() {
        if (ContextCompat.checkSelfPermission(this, permissions[0]) != 0 ||
                ContextCompat.checkSelfPermission(this, permissions[1]) != 0 ||
                ContextCompat.checkSelfPermission(this, permissions[2]) != 0) {
            ActivityCompat.requestPermissions(this, permissions, REQUEST_CODE)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_CODE && grantResults.isNotEmpty()) {
            grantResults.indices
                    .filter { grantResults[it] != PackageManager.PERMISSION_GRANTED }
                    .forEach { Log.d("PermissionCheckActivity", permissions[it] + " 没有授权") }
        }
    }

    companion object {
        protected val REQUEST_CODE = 1000
    }
}
